Declaring future meetup 2

trskop opened the pull request:
Declared future meetup 2 with presentation named "Types as values: Derive correctness from practicality".

While trying to understand the code I've made some notes and they were the basis for Haddock comments that are included in this pull request.

To simplify insertion of future meetups and to automatically generate meetup indexes, I've created default values for both Meetup and Presentation data types.

New "index.html" is not part of this pull request. I want to be sure that it will be accepted before doing so.

trskop committed:
Few comments that helped me to understand the code

trskop committed:
Generate meetup indexes; add future ones more simply

Data types Meetup and Presentation have sensible default values that can
be used to simplify definition of future meetups and presentations. As a
side effect of using such default values it makes more sense to generate
meetup indexes and in doing so prevent human errors when filling them
in.

trskop committed:
Declaring future meetup 2

Peter (me) will prepare presentation named:

    Types as values: Derive correctness from practicality

trskop committed:
Shell script for building index

Haskell executable "test-html.hs" has some dependencies that have to be
installed. Script "build.sh" installs them in Cabal sandbox and runs
"test-html.hs".

xkollar commented on gen/test-html.hs:
```diff
@@ -21,23 +22,68 @@ data Presentation = Presentation
-    , slides :: Maybe String
-    , audio :: Maybe String
-    , player :: Maybe String
+    , slides :: Maybe FilePath
```
While I see that `String` is probably not the best type for this, I'm not sure `FilePath` is better. It does not have to be relative path, it can also be absolute url to slides (even though storing presentation locally is preffered).

trskop replied:
How about declaring "type URL = String" for this purpose? Would you accept that?

xkollar commented on gen/test-html.hs:
```diff
@@ -21,23 +22,68 @@ data Presentation = Presentation
+
+-- | Assign index to a 'Meetup'. It expects them to be in reverse order, i.e.
+-- newest/future first and oldest as last.
+indexMeetups :: [Meetup] -> [Meetup]
```
I have also considered assigning index automatically, but decided against it as prevention of accidentally changing indexes in case of someone inserted meetup not on the right place. Maybe if there was a mechanism that would check whether meetups are ordered by date then I would be all for automatic indexing. Then, however, whole `indexM` would be next to useless.

trskop replied:
Ok, makes sense. How about switching to smart constructor for a future meetup? Something like:

``` Haskell
futureMeetup :: Integer -> Meetup
futureMeetup idx = Meetup
    { indexM = idx
    , presentations = []
    , time = Nothing
    , participants = Nothing
    }
```

That would also mean getting rid of "instance Default Meetup".

I would also change `indexMeetups` in to `checkMeetupIndexes` and instead of numbering them it would just check them and throw error if the indexes aren't correctly ordered. Can we agree on that?

From my experience people will most likely screw up small things. In this case using a particular number. Big things, like adding new meetup, are less likely to be done incorrectly and we have an advantage of code review. This discussion is a proof of that. :-)

xkollar commented:
Up to mentioned things I'm all up for the change. (I still plan move things to hakyll, but that will take some time `:-/`).

trskop committed:
Correcting printf formating string

Formating string "%n" is not valid for decimal integer, but "%d" is.

trskop committed:
Correcting Haddock comments; type URL = String

Attributes "slides", "audio", and "player" aren't relative paths, but
URLs. For this reason new type alias "type URL = String" was created to
help people in orienting by type signatures.

trskop committed:
Reverting back to manual meetup indexes

As a consequence there is no need for "instance Default Meetup" and
indexMeetups function. Instead we are using smart constructor:

    futureMeetup :: Integer -> Meetup

And also function that checks meetup index ordering:

    checkMeetupsIndex :: [Meetup] -> [Meetup]

Type signature doesn't reflect it, but that function throws exception
when meetups aren't ordered correctly. While partial functions aren't
good idea in complex applications, this usage provides help when trying
to add new Meetup incorrectly.

xkollar commented:
Thank you for incorporating my comments `:-)`. Why is pull request closed now?

xkollar commented:
Oh, I see... there is a new one `:-)`.
